unidist/core/backends/mpi/core/communication.py

```diff
@@ -116,39 +167,74 @@ def get_instance(cls, *args):
             cls.__instance = MPIState(*args)
         return cls.__instance
 
+    def is_root_process(self, rank=None):
+        """
+        Check if the rank is root process
 
-class MPIRank:
-    """Class that describes ranks assignment."""
+        Parameters
+        ----------
+        rank : int, default: the current rank
+            The rank to be checked
 
-    ROOT = 0
-    MONITOR = 1
-    FIRST_WORKER = 2
+        Returns
+        -------
+        bool
+            True or False.
+        """
+        if rank is None:
+            rank = self.global_rank
+        return rank == MPIRank.ROOT
 
+    def is_monitor_process(self, rank=None):
+        """
+        Check if the rank is monitor process
 
-def get_topology():
-    """
-    Get topology of MPI cluster.
+        Parameters
+        ----------
+        rank : int, default: the current rank
+            The rank to be checked
 
-    Returns
-    -------
-    dict
-        Dictionary, containing workers ranks assignments by IP-addresses in
-        the form: `{"node_ip0": [rank_2, rank_3, ...], "node_ip1": [rank_i, ...], ...}`.
-    """
-    mpi_state = MPIState.get_instance()
-    comm = mpi_state.comm
-    rank = mpi_state.rank
+        Returns
+        -------
+        bool
+            True or False.
+        """
+        if rank is None:
+            rank = self.global_rank
+        return rank in self.monitor_processes
 
-    hostname = socket.gethostname()
-    host = socket.gethostbyname(hostname)
-    cluster_info = comm.allgather((host, rank))
-    topology = defaultdict(list)
+    def get_monitor_by_worker_rank(self, global_rank=None):
+        """
+        Get the monitor process rank for the host that includes this rank.
 
-    for host, rank in cluster_info:
-        if rank not in [MPIRank.ROOT, MPIRank.MONITOR]:
-            topology[host].append(rank)
+        Parameters
+        ----------
+        global_rank : int
+            The global rank to search for a monitor process.
 
-    return dict(topology)
+        Returns
+        -------
+        int
+            Rank of monitor process.
+        """
+        if self.host_comm is None:
+            return MPIRank.MONITOR
+
+        if global_rank is None:
+            global_rank = self.global_rank
+        host = self.__host_by_rank[global_rank]
+        if host is None:
+            raise ValueError("Unknown rank of workers")
```

Comment:
When is it possible to get into this branch?

Reply:
It is just for safety in case this function is used incorectly.

Reply:
If there is no a host by a global rank, we will get an error on the line 224 (KeyError). If we still want to have this branch, then we should change the line 224 to the following.

```python
host = self.__host_by_rank.get(global_rank)
```

Reply:
No, __host_by_rank is defaultdict(None), so we can throw a user-readable exception if the host is None.